prediction-applier/src/main/java/org/alfresco/hxi_connector/prediction_applier/hx_insight/HxPredictionReceiver.java

```diff
@@ -0,0 +1,130 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
+package org.alfresco.hxi_connector.prediction_applier.hx_insight;
+
+import static org.apache.camel.LoggingLevel.DEBUG;
+import static org.apache.camel.LoggingLevel.TRACE;
+
+import java.util.List;
+import java.util.Objects;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.camel.Exchange;
+import org.apache.camel.Processor;
+import org.apache.camel.builder.RouteBuilder;
+import org.apache.camel.component.jackson.JacksonDataFormat;
+import org.apache.camel.component.jackson.ListJacksonDataFormat;
+import org.springframework.stereotype.Component;
+
+import org.alfresco.hxi_connector.common.model.prediction.Prediction;
+import org.alfresco.hxi_connector.prediction_applier.config.PredictionListenerConfig;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+@SuppressWarnings({"PMD.UnusedPrivateMethod", "PMD.UnusedFormalParameter", "PMD.LinguisticNaming", "PMD.LongVariable"})
+public class HxPredictionReceiver extends RouteBuilder
+{
+    private static final String PREDICTION_PROCESSOR_TRIGGER_ROUTE_ID = "prediction-processor-trigger-route";
+    private static final String PREDICTION_PROCESSOR_ROUTE_ID = "prediction-processor-route";
+    private static final String PREDICTION_PROCESSOR = "direct:prediction-processor";
+
+    private final PredictionListenerConfig config;
+
+    @Override
+    public void configure()
+    {
+        // @formatter:off
+        JacksonDataFormat predictionsBatchDataFormat = new ListJacksonDataFormat(Prediction.class);
+        JacksonDataFormat predictionDataFormat = new JacksonDataFormat(Prediction.class);
+
+        from(config.predictionProcessorTriggerEndpoint())
+                .routeId(PREDICTION_PROCESSOR_TRIGGER_ROUTE_ID)
+                .choice()
+                .when(this::isProcessingPending)
+                    .log(DEBUG, log, "Prediction processing is pending, no need to trigger it")
+                .otherwise()
+                    .log(DEBUG, log, "Triggering prediction processing")
+                    .to(PREDICTION_PROCESSOR);
+
+        from(PREDICTION_PROCESSOR)
+                .routeId(PREDICTION_PROCESSOR_ROUTE_ID)
+                .process(setIsProcessingPending(true))
+                .loopDoWhile(this::hasNextPage)
+                    .log(DEBUG, log, "Fetching predictions")
+                    .to(config.predictionsSourceEndpoint())
+                    .log(DEBUG, log, "Sending predictions to internal buffer: ${body}")
+                    .unmarshal(predictionsBatchDataFormat)
+                    .process(this::savePredictionsBatch)
+                    .loopDoWhile(this::predictionsBatchNotEmpty)
```

Comment:
I think it could be worth separating this loop to another route, WDYT?

Reply:
I was thinking about it but I came to the conclusion that it is better to keep it all in one place because both loops are working on the same body/variables/properties